optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/package-info.java

```diff
@@ -16,6 +16,50 @@
 
 /**
  * The {@link org.optaplanner.core.api.score.stream.ConstraintStream} API:
- * a way to define constraints for {@link org.optaplanner.core.api.score.Score} calculation.
+ * a way to define constraints for {@link org.optaplanner.core.api.score.Score} calculation. Throughout this
+ * documentation, we will be using the following terminology:
+ *
+ * <dl>
+ *     <dt>Stream</dt>
+ *          <dd>Short for "constraint stream". A chain of different operations, originated by
+ *          {@link org.optaplanner.core.api.score.stream.ConstraintFactory#from(java.lang.Class)} (or similar methods)
+ *          and terminated by a penalization or reward operation.</dd>
+ *     <dt>Operation</dt>
+ *          <dd>Operations (implementations of
+ *          {@link org.optaplanner.core.api.score.stream.ConstraintStream}) are parts of Stream which mutate it. They
+ *          may remove tuples from further evaluation, expand or contract streams. Every Stream has a terminal
+ *          Operation, which is either a penalization or a reward.</dd>
+ *     <dt>Fact</dt>
+ *          <dd>Object instance entering the Stream.</dd>
+ *     <dt>Genuine Fact</dt>
+ *          <dd>Fact that enters the Stream either through a from(...) call or through a join(...) call. Genuine Facts
+ *          are either planning entities (see {@link org.optaplanner.core.api.domain.entity.PlanningEntity}) or problem
+ *          facts (see @{@link org.optaplanner.core.api.domain.solution.drools.ProblemFactProperty} or
+ *          {@link org.optaplanner.core.api.domain.solution.drools.ProblemFactCollectionProperty}).</dd>
+ *     <dt>Inferred Fact</dt>
+ *          <dd>Fact that enters Stream through a computation. This would typically happen through an Operation such as
+ *          groupBy(...).</dd>
+ *     <dt>Tuple</dt>
+ *          <dd>A collection of Facts that Stream operates on. For example,
+ *          {@link org.optaplanner.core.api.score.stream.uni.UniConstraintStream} operates on single-Fact Tuples {A} and
+ *          {@link org.optaplanner.core.api.score.stream.bi.BiConstraintStream} operates on two-Fact Tuples {A, B}.
+ *          Putting Facts into a Tuple implies a relationship exists between these Facts.</dd>
```

Comment:
Things starting with a capital must be classes. Tuple and Fact are not classes, so let's call them tuple and fact (no capital start).